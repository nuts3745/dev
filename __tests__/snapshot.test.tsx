import { describe, it, expect } from "vitest";
import fs from "fs";
import path from "path";

describe("Snapshot Tests", () => {
  // wasm-game.jsのスナップショットテスト
  it("wasm-game.js file structure matches expected pattern", () => {
    const filePath = path.resolve("./scripts/wasm-game.js");
    const fileContent = fs.readFileSync(filePath, "utf8");

    // ファイル内容をフォーマットして一貫性のあるテスト結果を得る
    const normalizedContent = fileContent
      .replace(/\r\n/g, "\n") // 改行の正規化
      .replace(/[ \t]+$/gm, "") // 行末の空白を削除
      .trim(); // 先頭と末尾の空白を削除

    // スナップショットと比較
    expect(normalizedContent).toMatchSnapshot();
  });

  // wasm-game.tsxのスナップショットテスト
  it("wasm-game.tsx component matches expected pattern", () => {
    const filePath = path.resolve("./pages/wasm-game.tsx");
    const fileContent = fs.readFileSync(filePath, "utf8");

    // ファイル内容をフォーマットして一貫性のあるテスト結果を得る
    const normalizedContent = fileContent
      .replace(/\r\n/g, "\n") // 改行の正規化
      .replace(/[ \t]+$/gm, "") // 行末の空白を削除
      .trim(); // 先頭と末尾の空白を削除

    // スナップショットと比較
    expect(normalizedContent).toMatchSnapshot();
  });

  // next.config.jsのスナップショットテスト
  it("next.config.js configuration matches expected pattern", () => {
    const filePath = path.resolve("./next.config.js");
    const fileContent = fs.readFileSync(filePath, "utf8");

    // ファイル内容をフォーマットして一貫性のあるテスト結果を得る
    const normalizedContent = fileContent
      .replace(/\r\n/g, "\n") // 改行の正規化
      .replace(/[ \t]+$/gm, "") // 行末の空白を削除
      .trim(); // 先頭と末尾の空白を削除

    // スナップショットと比較
    expect(normalizedContent).toMatchSnapshot();
  });
});
