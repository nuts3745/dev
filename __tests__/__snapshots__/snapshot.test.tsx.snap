// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot Tests > next.config.js configuration matches expected pattern 1`] = `
"/** @type {import('next').NextConfig} */
module.exports = {
  reactStrictMode: true,
  output: "export",
  webpack: (config, { isServer }) => {
    config.experiments = {
      asyncWebAssembly: true,
      layers: true,
    };
    config.output.webassemblyModuleFilename = \`\${isServer ? "../" : ""}static/wasm/webassembly.wasm\`;
    return config;
  },
  images: {
    unoptimized: true,
  },
};"
`;

exports[`Snapshot Tests > wasm-game.js file structure matches expected pattern 1`] = `
"import {
  Cell,
  Universe,
} from "scripts/wasm-game-of-life/pkg/wasm_game_of_life";
import { memory } from "scripts/wasm-game-of-life/pkg/wasm_game_of_life_bg.wasm";

const defaultOptions = {
  cellSize: 5,
  gridColor: "#F8F7F2",
  deadColor: "#F8F7F2",
  aliveColor: "#333",
  sleepTime: 678,
};

const wasmImage = (id, options = defaultOptions) => {
  const cellSize = options.cellSize;
  const gridColor = options.gridColor;
  const deadColor = options.deadColor;
  const aliveColor = options.aliveColor;
  const sleepTime = options.sleepTime;

  const universe = Universe.new();
  const width = universe.width();
  const height = universe.height();

  const canvas = document.getElementById(id);
  canvas.height = (cellSize + 1) * height + 1;
  canvas.width = (cellSize + 1) * width + 1;

  const ctx = canvas.getContext("2d");

  let animationFrameId;

  const sleep = (waitTime) =>
    new Promise((resolve) => setTimeout(resolve, waitTime));
  const renderLoop = async () => {
    universe.tick();
    await sleep(sleepTime);
    drawGrid();
    drawCells();
    animationFrameId = requestAnimationFrame(renderLoop);
  };

  const drawGrid = () => {
    ctx.beginPath();
    ctx.strokeStyle = gridColor;

    for (let i = 0; i <= width; i++) {
      ctx.moveTo(i * (cellSize + 1) + 1, 0);
      ctx.lineTo(i * (cellSize + 1) + 1, (cellSize + 1) * height + 1);
    }

    for (let j = 0; j <= height; j++) {
      ctx.moveTo(0, j * (cellSize + 1) + 1);
      ctx.lineTo((cellSize + 1) * width + 1, j * (cellSize + 1) + 1);
    }

    ctx.stroke();
  };

  const getIndex = (row, column) => {
    return row * width + column;
  };

  const drawCells = async () => {
    const cellsPtr = universe.cells();
    const cells = new Uint8Array(memory.buffer, cellsPtr, width * height);

    ctx.beginPath();
    for (let row = 0; row < height; row++) {
      for (let col = 0; col < width; col++) {
        const idx = getIndex(row, col);

        ctx.fillStyle = cells[idx] === Cell.Dead ? deadColor : aliveColor;
        ctx.fillRect(
          col * (cellSize + 1) + 1,
          row * (cellSize + 1) + 1,
          cellSize,
          cellSize,
        );
      }
    }
    ctx.stroke();
  };

  // メモリを直接操作してセルの状態を切り替える
  const toggleCell = (row, col) => {
    const cellsPtr = universe.cells();
    const cells = new Uint8Array(memory.buffer, cellsPtr, width * height);
    const idx = getIndex(row, col);

    // セルの状態を反転（生きている→死、死んでいる→生）
    cells[idx] = cells[idx] === Cell.Dead ? Cell.Alive : Cell.Dead;

    // 変更をすぐに反映
    drawCells();
  };

  // キャンバス上のクリック位置からセルの行と列を計算
  const getCellPosition = (event) => {
    const boundingRect = canvas.getBoundingClientRect();

    const scaleX = canvas.width / boundingRect.width;
    const scaleY = canvas.height / boundingRect.height;

    const canvasLeft = (event.clientX - boundingRect.left) * scaleX;
    const canvasTop = (event.clientY - boundingRect.top) * scaleY;

    const row = Math.min(Math.floor(canvasTop / (cellSize + 1)), height - 1);
    const col = Math.min(Math.floor(canvasLeft / (cellSize + 1)), width - 1);

    return { row, col };
  };

  // クリックイベントハンドラ
  canvas.addEventListener("click", (event) => {
    const { row, col } = getCellPosition(event);
    toggleCell(row, col);
  });

  drawGrid();
  drawCells();
  requestAnimationFrame(renderLoop);
  return () => {
    if (animationFrameId) cancelAnimationFrame(animationFrameId);
  };
};

export default wasmImage;"
`;

exports[`Snapshot Tests > wasm-game.tsx component matches expected pattern 1`] = `
"import React, { useEffect } from "react";
import wasmImage from "scripts/wasm-game";

const WasmGame = () => {
  useEffect(() => {
    const cleanup = wasmImage("game-of-life");
    return cleanup;
  }, []);
  return (
    <div>
      <canvas id="game-of-life" />
    </div>
  );
};

export default WasmGame;"
`;
